fix: limit slack API calls

tinkerborg opened the pull request:
- only attempt to list/join channels if postMessage call fails with
  not_in_channel
- respect rate limit 429 responses when iterating through paginated
  conversations.list result

fixes #1206

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-events/blob/master/USERS.md).

<!--

Please leave your PR in draft if you don't need a review yet. 

To fix failing `CI / Codegen`, run `make codegen`. 

-->

whynowy commented:
@shashwat-appdirect @AleksanderGondek - Could you review this PR since both of you made code change on it? I think you have better knowledge than me.

tinkerborg commented:
I'm also fairly new to Go, so definitely open to suggestions for a better solution for the rate limiting here. ISTM that something like [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp) would implement it more generically for all the API methods here, but it wasn't clear for me, as a neophyte, how to utilize it here.

Apart from that, the other thing this PR does is implement a "try first, then ask for forgiveness" logic -- i.e., try to send the message and then try to join the channel only if that fails. The reasoning here is that the Slack API makes discovering a channel ID given its name for more complicated and expensive than it should be. Since the postMessage call doesn't require an ID, the ID is only needed during a join, so this logic defers that discovery until it's actually needed. Referring directly to the error text to determine this seems a bit brittle here, but I couldn't readily see another way to implement the logic.

shashwat-appdirect reviewed (approved):
The flow LGTM. Have you tested it both for public channel and private channel?

whynowy commented:
@shashwat-appdirect - could you help do a testing on the private channel stuff if @tinkerborg provides a testing image?

shashwat-appdirect commented:
> > The flow LGTM. Have you tested it both for public channel and private channel?
> 
> I did try this with a private channel, but it's not clear to me why/how the original code supported this. The private channel was not visible with my bot token and did not appear in the `conversations.list` output. The `chat.postMessage` call returns a `channel_not_found` error when trying to post to a private channel, rather than `not_in_channel`. What is the expected behavior here?

@tinkerborg you need to have proper permissions for your bot - channels:join, channels:read, chat:write, groups:read. Also, the slack app should be manually added to the private channel. Bot does not have permission to join a private channel.

shashwat-appdirect commented:
> @shashwat-appdirect - could you help do a testing on the private channel stuff if @tinkerborg provides a testing image?

Hi @whynowy , I think it will be easy for @tinkerborg to create a sensor image and test locally, since he will be having a setup. But still I could try to test.

tinkerborg commented:
I can test tomorrow morning.

On Tue, May 11, 2021 at 12:45 AM Shashwat Rawat ***@***.***>
wrote:

> @shashwat-appdirect <https://github.com/shashwat-appdirect> - could you
> help do a testing on the private channel stuff if @tinkerborg
> <https://github.com/tinkerborg> provides a testing image?
>
> Hi @whynowy <https://github.com/whynowy> , I think it will be easy for
> @tinkerborg <https://github.com/tinkerborg> to create a sensor image and
> test locally, since he will be having a setup. But still I could try to
> test.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/argoproj/argo-events/pull/1207#issuecomment-837794645>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADVJF6KHYDIYI3K2XSBIZKDTNCYLPANCNFSM44KYAHKA>
> .
>

shashwat-appdirect commented:
I am sure earlier `conversations.list` returned both public and private channels. To get private channel you need to pass `private_channel` in the field `types`, while calling `conversations.list` method. I tested just now using Slack API Tester - https://api.slack.com/methods/conversations.list/test and able to list all public and private channels.

tinkerborg commented:
> I am sure earlier `conversations.list` returned both public and private channels. To get private channel you need to pass `private_channel` in the field `types`, while calling `conversations.list` method. I tested just now using Slack API Tester - https://api.slack.com/methods/conversations.list/test and able to list all public and private channels.

That was my understanding as well, otherwise why would the option even exist? I'll do some digging and see if I can figure out what's going on.

tinkerborg commented:
@shashwat-appdirect Using my bot's token in the API tester, with `private_channel` argument, I can list private channels, but only the ones to which my bot has already been invited. So it's still not clear to me what the `isPrivateChannel` logic was doing.

shashwat-appdirect commented:
> @shashwat-appdirect Using my bot's token in the API tester, with `private_channel` argument, I can list private channels, but only the ones to which my bot has already been invited. So it's still not clear to me what the `isPrivateChannel` logic was doing.

`conversations.join` is not allowed for a private channel. Earlier if the channel was found in `conversations.list`, the `conversations.join` method was still called. But if the channel is private, this method fails with error - `method_not_supported_for_channel_type`. Hence, to avoid calling this method on private channels, `isPrivateChannel` check was used.
You could go through the old code flow, if still it is not clear :)

tinkerborg commented:
> > @shashwat-appdirect Using my bot's token in the API tester, with `private_channel` argument, I can list private channels, but only the ones to which my bot has already been invited. So it's still not clear to me what the `isPrivateChannel` logic was doing.
> 
> `conversations.join` is not allowed for a private channel. Earlier if the channel was found in `conversations.list`, the `conversations.join` method was still called. But if the channel is private, this method fails with error - `method_not_supported_for_channel_type`. Hence, to avoid calling this method on private channels, `isPrivateChannel` check was used.
> You could go through the old code flow, if still it is not clear :)

I didn't change this flow, just moved it to after trying `chat.postMessage`, and added the rate limiting. I don't see that it is doing anything though, since `conversations.list` only lists private channels if the bot is already joined. If you have a test case that the previous PR was addressing, I can try to reproduce and test with the code in this PR.

shashwat-appdirect commented:
Not any particular test case. I only tested if the message is posted in private channel. In your case you told that it is working after your code changes, so it should be fine. Don't think if anymore scenarios required.

whynowy commented on sensors/triggers/slack/slack.go:
```diff
@@ -114,54 +115,65 @@ func (t *SlackTrigger) Execute(ctx context.Context, events map[string]*v1alpha1.
+							break
+						}
+					}
+					if nextCursor == "" || channelID != "" {
```
👍

whynowy commented on sensors/triggers/slack/slack.go:
```diff
@@ -114,54 +115,65 @@ func (t *SlackTrigger) Execute(ctx context.Context, events map[string]*v1alpha1.
+				if channelID == "" {
+					return nil, errors.Errorf("failed to get channelID of %s", channel)
+				}
+				if isPrivateChannel {
```
I actually not very sure what the logic for private channel should be, but here the logic seems to be changed.

Old: Skip "join conversation", still post message;
New: return an error.

tinkerborg replied:
The error probably makes sense IMO, because if it's a private channel the bot can't join and can't send a message to it without joining. It needs to be invited.

whynowy replied:
Sounds good.

whynowy commented on sensors/triggers/slack/slack.go:
```diff
@@ -114,54 +115,65 @@ func (t *SlackTrigger) Execute(ctx context.Context, events map[string]*v1alpha1.
+	t.Logger.Infow("posting to channel...", zap.Any("channelName", channel))
 	for {
-		channels, nextCursor, err := api.GetConversations(params)
+		channelID, timestamp, err := api.PostMessage(channel, slack.MsgOptionText(message, false))
```
With the code change, do we still need the for loop in line 119? I seems to be infinite.

tinkerborg replied:
This is to implement retry logic, i.e. try to post message first, then if it fails due to not_in_channel error, join the channel and try again.

It shouldn't be infinite, as all paths of logic inside the loop lead to a return, either due to success or error. If the `api.JoinConversation` call on line 161 incorrectly returns a `nil` value for `err` when it actually did fail under the hood, this could loop infinitely. Not sure how likely that is, but I could limit the loop to max 2 iterations and add some sort of "unspecified error" logging after it.

whynowy replied:
I see.

whynowy reviewed (approved):
LGTM, thanks!
